### ISSUE TITLE
Is there a way to stream test results to the xharness client?

### ISSUE THREAD

**Issue by borrrden** [dotnet/xharness#994]
I'm in the middle of transitioning to this tool and currently when I am testing either with CLI xunit, or a similar solution to xharness that I dug up from the mono repo circa 2011 that streams results over a TcpWriter like this, I can see a list of tests go by as they pass and fail.  I'm trying to find a way to do this using xharness, but even with maximum logging it seems to simply log "Test Execution started" and then nothing after that while the tests are running.  I am currently attempting this via Mac Catalyst, and the results that I see in the output directory are a large app specific log full of messages from the library I am testing, and MacCatalyst.system.log which is size 0, and then test-maccatalyst.log which is also size 0.  My tests seem to have a native crash inside of them and this manifests as a hang with no other information present.  

So my question / request would be is there a way that xharness can print out tests as it goes (I think the README says that tests starting is implemented for WASM only, but no mention of test ending).  I also can't seem to find a way from the test harness side to send messages to xharness to be displayed (I could use the events to send that information if there were a way).

**Comment by premun** [comment]
Hey @borrrden,
This is something [we've been thinking about](https://github.com/dotnet/xharness/issues/391) but it would require a substantial amount of work. Currently, the data we're sending are really not structured in any way and we only expect the contents of the test results. Instead, we'd need to start sending structured data (some messages) and parse and process them in XHarness. This means quite an overhaul of the TCP connection which would probably introduce some complexity (and bugs). Considering that XHarness is primarily used in CI where we wouldn't take advantage of this, it didn't get funded. XHarness is quite stable these days (last year) and powers a substantial part of .NET's infrastructure so I think it would be hard to fund this as it only brings risk and no value.

Now to your issue - I think it's interesting that MacOS doesn't detect the catalyst app ending and return automatically from the process call. We don't see hangs like that really but rather the `APP_CRASH` exit code.

What would be possible though, and I think it should be quite easy to do, is to add a logger that would scan the system log (like we already do) and relay the log into console in real time. If this behaviour was behind a switch (e.g. `--live-logs` or similar), I can even see having this checked in here (as opposed to you forking the repo and adding the functionality for instance). However, this would have to be a contribution on your part. I could definitely help point you in the right direction though.

**Comment by borrrden** [comment]
Thanks for the response.  It's interesting that it went in this sort of direction.  I'm trying to transition from Xamarin to .NET 6, and previously we have used [the xUnit device runner](https://github.com/xunit/devices.xunit) combined with the aforementioned TCP reception server [committed here in our repo](https://github.com/couchbase/couchbase-lite-net/blob/master/Tools/Touch.Server/Main.cs) to stream what xunit usually logs (i.e. pass pass pass pass fail fail) and scan for the final expected line (passed: X failed: Y skipped: Z or something like that).  We then parse the results from there.  I imagine this is similar to what VSTest does when running UWP / WinUI tests.  

So if I understand correctly, this program doesn't actually transmit results in realtime, but rather just opens up a few logging sinks into files that relay information that is accessible in other ways (idevicesyslog, adb log, log show, xharness logs from itself) and then waits for a TCP dump with the final results of the entire run?  That's a bit unfortunate but I believe I should be able to mimic the current behavior by keeping around the 2011 Touch.Server and opening up a separate TCP connection with hooks into the TestStarted / TestEnded events.  I was really hoping that I could retire all that stuff though.  Maybe I could merge that functionality into a second tcp stream inside a fork of xharness though.

We also use this heavily for CI, but it's just useful to be able to scan the logs of a failed job on Jenkins to see which test failed, or if impatient we can just watch the tests scroll by to confirm it's not stuck.

**Comment by premun** [comment]
Now I'm thinking - if you only care about maccatalyst, you should be fine with reading the system log and filtering by the app name. No need for any TCP dance.

**Comment by premun** [comment]
- Launching MacCatalyst - https://github.com/dotnet/xharness/blob/b98a9186700bd659b156f89439ee17a8c72d2d0b/src/Microsoft.DotNet.XHarness.Apple/AppOperations/AppRunnerBase.cs#L106-L114
- Capturing system log of an app - https://github.com/dotnet/xharness/blob/b98a9186700bd659b156f89439ee17a8c72d2d0b/src/Microsoft.DotNet.XHarness.Apple/AppOperations/AppRunnerBase.cs#L304-L338

**Comment by premun** [comment]
It should be trivially easy to append a `CallbackLog` onto this `log` using `AggregatedLog` and make `CallbackLog` log to the console everything.

**Comment by borrrden** [comment]
Thanks for this.  I am going to be using xharness for .NET 6 iOS Android and Mac Catalyst (and maybe WinUI, but for now that is a separate test project) so it is not just Mac Catalyst.  You are correct that the result lines alone are not helpful for knowing which test is hanging.  I would hope that I could also print which tests are starting as well.  Anyway it sounds like a project for a fork.

I'm reasonably sure that simply running the app hangs as well so it might be something to do with mac apps in general, and not specific to xharness, but killing the app I think should be seen by xharness (at least via a closed TCP stream).  I am unfamiliar with this architecture since we didn't support Xamarin Mac and are only coming into mac catalyst now in .NET 6 (whereas previously we supported Xamarin iOS and Xamarin Android).

**Comment by premun** [comment]
@borrrden please have a look at the `--launch-timeout` and `--timeout` options which should help you kill the run. The former waits for the TCP connection, the latter applies to the duration of the whole run.

Secondly, now sure how you're running the tests, but if through xUnit, there is this "output long running tests option" that you configure through the xunit runner JSON file. It can then write a line each X seconds a test is running. This helped us in our infra previously. That is, when you actually get the output back of course.

**Comment by borrrden** [comment]
Ok I will definitely have a look at those.  I actually finally figured out how to turn up the logging level so that test names get output into the main log (log show).  It seems like the beforeteststartingevent never fires on mac catalyst, and so I never get the TestStarted event, though I do get the TestEnded event.  The test harness itself is able to log "Test Starting" and such though so I found the test that was causing the issue and fixed it.  I'm unfamiliar with the xunit runner json file, that's something I'll look into as well.  Thanks for all the advice.  Testing a bazillion different platforms can be frustrating.
